addons/addon-base-raas/packages/base-raas-services/lib/budgets/budgets-service.js

```diff
@@ -0,0 +1,255 @@
+    // Update thresholds
+    requestParams = {
+      AccountId: accountId,
+      BudgetName: this.budgetName,
+    };
+    const budgetApiNotificationResponse = await budgetsClient.describeNotificationsForBudget(requestParams).promise();
+    const notifications = budgetApiNotificationResponse.Notifications;
+    const oldThresholds = notifications.map(x => x.Threshold);
+    const newThresholds = _.get(newBudgetConfiguration, 'thresholds', []);
+    // Check if thresholds need update
+    if (_.isEqual(oldThresholds.sort(), newThresholds.sort())) {
+      // Check if there's notifications
+      if (notifications.length === 0) {
+        return;
+      }
+      // Check if notification email needs update
+      // spot check one email here as Galileo always use the same email for all notification
+      // Assume no one is updating the Budget info from AWS console now
+      const oldNotificationEmail = this._getANotificationEmail(requestParams, notifications[0], budgetsClient);
+      if (oldNotificationEmail === newBudgetConfiguration.notificationEmail) {
+        return;
+      }
+    }
+    await this._updateNotificationsNSubscribers(notifications, newBudgetConfiguration, budgetsClient, requestParams);
+  }
+
+  async _createBudget(accountId, budgetsClient, newBudgetConfiguration) {
+    const requestParams = {
+      AccountId: accountId,
+      Budget: this._formBudgetObject(newBudgetConfiguration),
+      NotificationsWithSubscribers: this._formNotificationObjects(newBudgetConfiguration),
+    };
+    await this._runAndCatchInvalidParameterError(async () => budgetsClient.createBudget(requestParams).promise());
+  }
+
+  async _getANotificationEmail(requestParams, notification, budgetsClient) {
+    const requestParamsClone = _.clone(requestParams);
+    requestParamsClone.Notification = notification;
+    const budgetApiSubscribersResponse = await budgetsClient
+      .describeSubscribersForNotification(requestParamsClone)
+      .promise();
+    return budgetApiSubscribersResponse.Subscribers[0].Address;
+  }
+
+  async _updateNotificationsNSubscribers(oldNotifications, newBudgetConfiguration, budgetsClient, requestParams) {
+    // Delete existing notifications
+    const deletePromises = oldNotifications.map(notification =>
+      budgetsClient.deleteNotification({ ...requestParams, Notification: notification }).promise(),
+    );
+    await Promise.all(deletePromises);
+    // Add new notifications
+    if (_.has(newBudgetConfiguration, 'thresholds')) {
+      const newNotifications = this._formNotificationObjects(newBudgetConfiguration);
+      const createPromises = newNotifications.map(notification =>
+        budgetsClient.createNotification({ ...requestParams, ...notification }).promise(),
+      );
+      await this._runAndCatchInvalidParameterError(async () => Promise.all(createPromises));
+    }
+  }
+
+  _formBudgetObject(newBudgetConfig) {
+    const budget = this.budgetBaseTemplates.BudgetTemplate;
```

Comment:
I think you need to do _.cloneDeep() on this otherwise, race conditions will produce strange behavior for admins when they access the UI.

Reply:
Good catch! Will send a PR shortly.